pkg/testing/tools/artifacts_snapshot.go

```diff
@@ -0,0 +1,204 @@
+package tools
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/google/uuid"
+	"golang.org/x/exp/maps"
+)
+
+const (
+	defaultAPIURL         = "https://artifacts-snapshot.elastic.co"
+	beatsManifestEndpoint = "beats/latest/%s.json"
+)
+
+type ArtifactSnapshotClient struct {
+	url        string
+	httpClient *http.Client
+}
+
+type ArtifactsManifestResponse struct {
+	BuildID     string `json:"build_id"`     // example value: "8.8.3-b1d8691a"
+	Version     string `json:"version"`      // example value: "8.8.3-SNAPSHOT"
+	ManifestURL string `json:"manifest_url"` // example value: https://artifacts-snapshot.elastic.co/beats/8.8.3-b1d8691a/manifest-8.8.3-SNAPSHOT.json
+	SummaryURL  string `json:"summary_url"`  // example value: https://artifacts-snapshot.elastic.co/beats/8.8.3-b1d8691a/summary-8.8.3-SNAPSHOT.html
+}
+
+type Projects struct {
+	Projects map[string]ProjectData `json:"projects"`
+}
+
+type ProjectData struct {
+	Branch     string                 `json:"branch"`
+	CommitHash string                 `json:"commit_hash"`
+	CommitURL  string                 `json:"commit_url"`
+	Packages   map[string]PackageData `json:"packages"`
+}
+
+type PackageData struct {
+	Url          string            `json:"url"`
+	ShaUrl       string            `json:"sha_url"`
+	Type         string            `json:"type"`
+	Architecture string            `json:"architecture"`
+	OS           []string          `json:"os"`
+	Attributes   map[string]string `json:"attributes"`
+}
+
+type PackageRequest struct {
+	// heartbeat-8.10.0-SNAPSHOT-linux-x86_64.tar.gz
+	Name       string
+	TargetPath string
+}
+
+func NewArtifactSnapshotClient() *ArtifactSnapshotClient {
+	return &ArtifactSnapshotClient{
+		url:        defaultAPIURL,
+		httpClient: new(http.Client),
+	}
+}
+
+func (c *ArtifactSnapshotClient) DownloadPackages(ctx context.Context, packageRequests []PackageRequest, version string) error {
+	manResponse, err := c.getManifestResponse(ctx, version)
+	if err != nil {
+		return fmt.Errorf("getting manifest response: %w", err)
+	}
+	manifestPackages, err := c.getManifestPackages(ctx, manResponse.ManifestURL)
+	if err != nil {
+		return fmt.Errorf("getting manifest packages: %w", err)
+	}
+	for _, pkg := range packageRequests {
+		pkgData, ok := manifestPackages[pkg.Name]
+		if !ok {
+			return fmt.Errorf("package %s not found in manifest", pkg)
+		}
+		err := c.downloadPackage(ctx, pkgData, pkg.TargetPath)
```

Comment:
All the packages can be downloaded concurrently. There is now an example using errgroup in the magefile on main.